src/actions/groups.cr

```diff
@@ -1,34 +1,176 @@
 module DppmRestApi::Actions::Groups
+  class NoSuchGroup < NotFound
+    def initialize(context : HTTP::Server::Context, id : String, cause : Exception? = nil)
+      super context, "No group found with ID " + id, cause
+    end
+  end
+
+  class DuplicateGroup < InternalServerError
+    def initialize(context, id, cause : Exception? = nil)
+      super context, "Two groups found with ID #{id}!", cause
+    end
+  end
+
+  class InvalidAccessLevel < BadRequest
+    def initialize(context : HTTP::Server::Context, cause : Exception? = nil)
+      level = context.params.url["access_level"]? || context.params.query["access_level"]?
+      super context, %[Given access level "#{level}" is not valid], cause
+    end
+  end
+
+  def query_params_in_body?(of context : HTTP::Server::Context) : Hash(String, Array(String))
+    if body = context.request.body
+      Hash(String, Array(String)).from_json body
+    else
+      Hash(String, Array(String)).new
+    end
+  rescue e : JSON::ParseException
+    raise BadRequest.new context, <<-HERE, cause: e
+      Expected either a valid JSON Hash of String to Array of String, or no
+      specified request body at #{context.request.path}
+      HERE
+  end
+
   extend self
   include RouteHelpers
   relative_post do |context|
     raise Unauthorized.new context unless Actions.has_access? context, Access::Create
-    # TODO: create a group based on the response body
-
+    body = context.request.body
+    raise BadRequest.new context, "One must specify a group to add." if body.nil?
+    DppmRestApi.permissions_config.groups << Config::Group.from_json body
+    DppmRestApi.permissions_config.sync_to_disk
   end
-  relative_put "/:id/route" do |context|
-    raise Unauthorized.new context unless Actions.has_access? context, Access::Update
-    # TODO: add a new route to the group
 
-  end
-  relative_put "/:id/param" do |context|
+  # Regardless of whether the given path exists on the given group, set the
+  # group's access level on that path to the given access level.
+  relative_put "/:id/route/:path/:access_level" do |context|
     raise Unauthorized.new context unless Actions.has_access? context, Access::Update
-    # TODO: add a new param to the group
-
+    group_id_s = context.params.url["id"]
+    group_id = group_id_s.to_i? || raise BadRequest.new context,
+      %[Group ID #{group_id_s} was not a valid integer]
+    group_already_found = false
+    path = URI.unescape context.params.url["path"]
+    access_level = Access.parse?(context.params.url["access_level"])
+    raise InvalidAccessLevel.new context if access_level.nil?
+    DppmRestApi.permissions_config.groups.map! do |group|
```

Comment:
Why recreating an Array?
We just need to get the permissions Hash of a group, then modifying this reference.